packages/api/src/WithServerEvents.tsx

```diff
@@ -0,0 +1,241 @@
+/* tslint:disable */
+import * as React from 'react';
+import { IHeaders } from '@syndesis/api';
+import { resolve as _resolve } from 'url';
+
+export interface IChangeEvent {
+  action: string;
+  kind: string;
+  id: string;
+}
+
+export interface IMessageEvent {
+  id: string;
+  data: string;
+  event: string;
+}
+
+export interface IWithEventsProps {
+  apiUri: string;
+  headers: IHeaders;
+  children(props: IWithEventsState): any;
+}
+
+export interface IWithEventsState {
+  registerChangeListener: (
+    key: string,
+    listener: (event: IChangeEvent) => void
+  ) => void;
+  registerMessageListener: (
+    key: string,
+    listener: (event: IMessageEvent) => void
+  ) => void;
+  unregisterChangeListener: (key: string) => void;
+  unregisterMessageListener: (key: string) => void;
+}
+
+const RECONNECT_TIME = 5000;
+
+export class WithServerEvents extends React.Component<
+  IWithEventsProps,
+  IWithEventsState
+> {
+  private starting = false;
+  private started = false;
+  private retries = 0;
+  private preferredProtocol: string = '';
+  private eventSource: EventSource | undefined;
+  private webSocket: WebSocket | undefined;
+  private apiEndpoint: string | undefined;
+  private headers: IHeaders | undefined;
+  private changeListeners = {};
+  private messageListeners = {};
+
+  public constructor(props: IWithEventsProps) {
+    super(props);
+    this.state = {
+      registerChangeListener: this.registerChangeListener.bind(this),
+      registerMessageListener: this.registerChangeListener.bind(this),
+      unregisterChangeListener: this.unregisterChangeListener.bind(this),
+      unregisterMessageListener: this.unregisterMessageListener.bind(this),
+    };
```

Comment:
As above, bind the functions in the constructor body and don't set them in the state unless absolutely necessary.

Reply:
Awesome, also spotted a copypasta error I made too, phew!